_stbt/core.py

```diff
@@ -159,8 +159,22 @@ def __exit__(self, exc_type, exc_value, tb):
             self._sink_pipeline.__exit__(exc_type, exc_value, tb)
             self._sink_pipeline = None
             self._mainloop.__exit__(exc_type, exc_value, tb)
+            self._running = False
         self._control = None
 
+    def set_sink_pipeline(self, new_sink_pipeline):
+        old_sink_pipeline = self._sink_pipeline
+        assert isinstance(self._sink_pipeline, NoSinkPipeline)
+        if self._running:
+            new_sink_pipeline.__enter__()
+        try:
+            self._sink_pipeline = new_sink_pipeline
+            self._display.set_sink_pipeline(new_sink_pipeline)
+        finally:
+            if self._running:
+                old_sink_pipeline.exit_prep()
+                old_sink_pipeline.__exit__(None, None, None)
```

Comment:
Really this is only safe if `old_sink_pipeline` is `NoSinkPipeline`.  `Display.on_new_sample` calls `sink_pipeline.on_sample`, and may call it on the old sink pipeline after or during `old_sink_pipeline.__exit__`.  In practice I don't think this will be a problem.